### PR TITLE
feat(frontend): push the calculation of inputs of join's condition to its two sides.

### LINKED ISSUE #3478: bug: panic when two sides of `EqJoinPredicate` have different types
**Describe the bug**
For `a = b`, we allow `a` and `b` to have different types, as long as they are castable to a more general type. This "cast" happens during evaluation of the `Equal` expression directly, without frontend binder inserting a cast expr.

However, this also makes **HashJoin** more complicated. It seems we have been assuming they are always the same type and may produce wrong result, runtime error, or even panic when they are not.

**To Reproduce**
```
-- batch: bigint works well
with a(a1) as (values (1)), b(b1) as (values (1::bigint)) select * from a, b where a1 = b1;
-- batch: decimal outputs wrong empty result
with a(a1) as (values (1)), b(b1) as (values (1::numeric)) select * from a, b where a1 = b1;
-- batch: smallint causes runtime error
with a(a1) as (values (1)), b(b1) as (values (1::numeric)) select * from a, b where a1 = b1;

-- streaming: panic
create table a(a1 int); create table b(b1 bigint); create table c(c1 smallint);
create materialized view big as select * from a, b where a1 = b1;
insert into a values (1);
```

**Expected behavior**
```
 a1 | b1 
----+----
  1 |  1
(1 row)
```

**Additional context**

* May also impact #3419. This is actually inspired by Question 2 in the issue.
* [PostgreSQL](https://github.com/postgres/postgres/blob/7201cd18627afc64850537806da7f22150d1a83b/src/include/catalog/pg_operator.dat#L20-L24) also allows direct cross-type comparison without cast exprs. And it is actually mixed style: sometimes it compares directly and sometimes it relies on implicit cast expr. In RW we always compare directly as of now.

Candidate solutions:
* HashJoin and DynamicFilter also implement direct cross-type comparison.
* Disallow direct cross-type comparison completely, and always insert implicit cast expr in frontend.
* Disallow direct cross-type equality inside `EqJoinPredicate`. Implicit cast expr should be inserted when creating `EqJoinPredicate`. However, `EqJoinPredicate` only accepts `InputRef` and it would need an extra `Project` to do the casting. Extending `EqJoinPredicate` to accept `ExprImpl` may also complicate other things.
* `EqJoinPredicate` (and `DynamicFilter`) carries the `DataType` that both sides should be casted to, and does the cast before using values. For example:
```diff
pub struct EqJoinPredicate {
    other_cond: Condition,

-   eq_keys: Vec<(InputRef, InputRef)>,
+   eq_keys: Vec<(InputRef, InputRef, DataType)>,

    left_cols_num: usize,
}
```

### PR DESCRIPTION
**Pull request by likg227**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
  - push the calculation of inputs of join's comparison condition to its two sides
  - fix a bug of join.col_prune
- How does this PR work? Need a brief introduction for the changed logic (optional)
It adds two project for join's two inputs to do the calculation.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

close #3478 #3934 #3713

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 1]
```diff
@@ -433,6 +434,90 @@ impl LogicalJoin {
...
+                    continue;
+                };
+                // when both `left` and `right` are `input_ref`, and they have the same return type or `func` is non-equal comparison expression, there is no need to calculate them in project.
+                if left.as_input_ref().is_some() && right.as_input_ref().is_some() && (left.return_type() == right.return_type() || !can_cast) {
+                    continue;
+                }
+                if can_cast {
+                    // align return types to avoid error when executing join.
+                    let mut v = vec![left, right];
+                    let _data_type = align_types(v.iter_mut()).unwrap();
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:489]:
```suggestion
                    align_types(v.iter_mut()).unwrap();
```

**Review by st1page — APPROVED** [review]
generally LGTM, btw I think some local variable could be immutable which make code more clear(just personal preferences).

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 2]
```diff
@@ -433,6 +434,90 @@ impl LogicalJoin {
...
+                ty,
+                Type::LessThan
+                    | Type::LessThanOrEqual
+                    | Type::Equal
+                    | Type::GreaterThan
+                    | Type::GreaterThanOrEqual
+            )
+        };
+        for (index, expr) in exprs.iter().enumerate() {
+            if let ExprImpl::FunctionCall(func) = expr && is_comparison_type(func.get_expr_type()) {
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:462]:
```
let ExprImpl::FunctionCall(func) = expr else { continue };
if !is_comparison_type(func.get_expr_type()) {
    continue;
}
```
to have one less level of indentation.

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:489]: A simpler way:
```
align_types([&mut left, &mut right].into_iter()).unwrap();
```
This avoids intermediate `v` and `next_tuple` as well.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 3]
```diff
@@ -597,6 +682,11 @@ impl_plan_tree_node_for_binary! { LogicalJoin }
 
 impl ColPrunable for LogicalJoin {
     fn prune_col(&self, required_cols: &[usize]) -> PlanRef {
+        // make `required_cols` point to internal table instead of output schema.
+        let required_cols = required_cols
+            .iter()
+            .map(|i| self.output_indices[*i])
+            .collect_vec();
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:689]: As discussed offline, let's try to make this bug fix a separate PR with test cases.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 4]
```diff
@@ -702,26 +788,94 @@ impl PredicatePushdown for LogicalJoin {
...
+            // the internal table of join has has the following schema:
+            // original left's columns | left_exprs | original right's columns | right_exprs
+            //```
+            // `left_index` and `right_index` will scan through `left_exprs` and `right_exprs`
+            // respectively.
+            let mut left_index = left_col_num;
+            let mut right_index = left_col_num + left_exprs.len() + right_col_num;
+            // replace chosen function calls.
+            for (i, (index_of_func_call, ty)) in
+                indices_and_ty_of_func_calls.into_iter().enumerate()
```
Comment (xiangjinwu) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:843]:
```
for ((left_expr, right_expr), (index_of_func_call, ty)) in left_exprs
    .iter()
    .zip_eq(&right_exprs)
    .zip_eq(indices_and_ty_of_func_calls)
```
to avoid indexing `left_exprs[i]`

**Review by xiangjinwu — COMMENTED** [review]
Rest lgtm. As discussed offline, this can be a standalone optimizer rule rather than part of predicate pushdown.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 5]
```diff
@@ -702,26 +788,94 @@ impl PredicatePushdown for LogicalJoin {
             LogicalJoin::can_push_on_from_filter(join_type),
         );
 
-        let mut new_on = self.on.clone().and(on);
-        let (left_from_on, right_from_on, on) = LogicalJoin::push_down(
-            &mut new_on,
+        let mut on = self.on.clone().and(on);
+        let (left_from_on, right_from_on, old_on) = LogicalJoin::push_down(
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:792]: just a nit, `old_on` could be better understood if named `on_from_on`

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 6]
```diff
@@ -702,26 +788,94 @@ impl PredicatePushdown for LogicalJoin {
...
+                    .collect_vec();
+                ColIndexMapping::new(map)
+            };
+            let (mut exprs, new_output_indices) =
+                Self::remap_exprs_and_output_indices(exprs, output_indices, &mut col_index_mapping);
+            output_indices = new_output_indices;
+
+            // ```ignore
+            // the internal table of join has has the following schema:
+            // original left's columns | left_exprs | original right's columns | right_exprs
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:835]: Will we then prune away the unneeded columns in "original left's columns" and "original right's columns" after pushing down the expressions?

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 7]
```diff
@@ -702,26 +788,94 @@ impl PredicatePushdown for LogicalJoin {
...
+                    .schema()
+                    .data_types()
+                    .into_iter()
+                    .enumerate()
+                    .map(|(i, data_type)| InputRef::new(i, data_type).into())
+                    .collect_vec();
+                exprs.extend(appended_exprs);
+                LogicalProject::create(input, exprs)
+            };
+            if !left_exprs.is_empty() {
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:870]: The right_exprs is_empty case should be handled separately?

File: src/frontend/test_runner/tests/testdata/join.yaml [hunk 8]
```diff
@@ -275,3 +275,33 @@
...
+        LogicalScan { table: t2, columns: [x, y] }
+- sql: |
+    create table t1(x int, y int);
+    create table t2(x float, y int);
+    select * from t1, t2 where t1.x + t1.y = t2.x;
+  optimized_logical_plan: |
+    LogicalJoin { type: Inner, on: ((t1.x + t1.y)::Float64 = t2.x) }
+      LogicalProject { exprs: [t1.x, t1.y, (t1.x + t1.y)::Float64] }
+        LogicalScan { table: t1, columns: [x, y] }
+      LogicalProject { exprs: [t2.x, t2.y, t2.x] }
```
Comment (jon-chuang) [inline src/frontend/test_runner/tests/testdata/join.yaml:296]: This is an unnecessary project created by the bug pointed out above.

File: src/frontend/test_runner/tests/testdata/nexmark.yaml [hunk 9]
```diff
@@ -272,9 +272,10 @@
         BatchFilter { predicate: (bid.date_time >= ((TumbleStart(bid.date_time, '00:00:10':Interval) + '00:00:10':Interval) - '00:00:10':Interval)) AND (bid.date_time <= (TumbleStart(bid.date_time, '00:00:10':Interval) + '00:00:10':Interval)) }
           BatchHashJoin { type: Inner, predicate: bid.price = max(bid.price) }
             BatchExchange { order: [], dist: HashShard(bid.price) }
-              BatchScan { table: bid, columns: [auction, bidder, price, date_time] }
+              BatchProject { exprs: [bid.auction, bid.bidder, bid.price, bid.date_time, bid.date_time] }
```
Comment (jon-chuang) [inline src/frontend/test_runner/tests/testdata/nexmark.yaml:275]: Another unnecessary project

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml [hunk 10]
```diff
@@ -185,8 +185,9 @@
             LogicalScan { table: t2, columns: [_row_id, x, y] }
   optimized_logical_plan: |
     LogicalJoin { type: LeftSemi, on: (t1.y = t2.y) AND (t1.x > (t2.x + 1000:Int32)) }
-      LogicalScan { table: t1, columns: [x, y] }
-      LogicalProject { exprs: [t2.y, t2.x] }
+      LogicalProject { exprs: [t1.x, t1.y, t1.x] }
```
Comment (jon-chuang) [inline src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:188]: Another unnecessary project.

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 6 above)
Comment (likg227) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:835]: Actually, this is the schema of join's internal table, which consists of left and right's output schema. So if you mean whether left's schema will be pruned, the answer is yes.

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 7 above)
Comment (likg227) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:870]: no. because left_exprs and right_exprs are of the same length, we just need to check left_exprs.is_empty() here.

File: src/frontend/test_runner/tests/testdata/join.yaml (hunk 8 above)
Comment (likg227) [reply src/frontend/test_runner/tests/testdata/join.yaml:296]: Actually, this is an unnecessary project but not a bug. we just append left_exprs and right_exprs to left project and right project respectively, so the condition of join refers to them. We can remove the right project in this case, but we just keep it for simplicity now.

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/join.yaml:296]: Ok, so we will depend on some other rule (e.g. project join merge, project scan merge) to remove the project afterwards?

Comment (likg227) [reply src/frontend/test_runner/tests/testdata/join.yaml:296]: If we want to remove this project, I think we should change the indices of LogicalJoin's `on`.

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/join.yaml:296]: Why can't we just make things simpler and don't introduce unnecessary project? It doesn't seem like we need to always symmetrically change LHS and RHS. The LHS can have no exprs etc.

Comment (xiangjinwu) [reply src/frontend/test_runner/tests/testdata/join.yaml:296]: > Why can't we just make things simpler and don't introduce unnecessary project

We can definitely make the plan simpler, at the cost of making the code logic more complicated. It is still worth it, but not necessarily in the first PR that enables this feature. It is easier to review if we do less in each PR.

**Review by xiangjinwu — APPROVED** [review]
